### PR TITLE
Correct distribution of DOA to primary air inlet nodes of ZoneHVAC

### PR DESCRIPTION
**Pull request by Nigusse**
Issue  #6020

***This description needs to change to match the final solution***
Now the OA is correctly distributed to air terminal mixer and zones per the amount specified in OA Design Specification object. This issue has been addressed for FanCoil, WSHP, PTAC, PTHP and VRF units. 

Pull request overview
---------------------
This is a defect. Diffs are expected because, now a different amount of OA is used instead of zero. Diffs are expected in EIO, audit and csv output files. Idd has been cleaned. Notes has been removed from OA input fields for each object.   And I/O reference document has also been updated.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully

### CODE REVIEW

**Comment by Nigusse** [comment]
Now DOA is correctly distributed to the terminal air mixers connected to the air loop per the zone outdoor air design specification. Example files that use terminal air mixer objects are expected to show diffs. Some unit test files that failed due to this fix has also been addressed. Unit tests are going to be added.

**Comment by Nigusse** [comment]
Added CO2 and general contaminant update equation to air terminal mixer model.

**Comment by nrel-bot-3** [comment]
@Nigusse @lgentile it has been 15 days since this pull request was last updated.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc
Comment (Myoldmopar) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc]: This big comment should probably just be removed right?

File: src/EnergyPlus/PackagedTerminalHeatPump.cc
Comment (Myoldmopar) [inline src/EnergyPlus/PackagedTerminalHeatPump.cc]: It's great if this isn't needed, just go ahead and remove it.

File: testfiles/DOAToFanCoilInlet.idf
Comment (Myoldmopar) [inline testfiles/DOAToFanCoilInlet.idf:2130]: Transition rule needed?

File: testfiles/DOAToPTHP.idf
Comment (Myoldmopar) [inline testfiles/DOAToPTHP.idf:160]: Not sure if the black invalid character or the plain question mark is better here.  Actually, leaving the bad character in there is a better option because it's easier to search and replace later.  I can try to undo these changes to the files if you need help @lgu1234, since I assume you didn't manually make this change, your editor did.

File: src/EnergyPlus/SingleDuct.cc
Comment (Myoldmopar) [inline src/EnergyPlus/SingleDuct.cc]: Moving forward, it's fine to just define the variables on the same line as the assignment:

```
int PriInNode = SysATMixer( SysNum ).PriInNode;
```

No big deal, just feel free to do that in the future.

There is, however, this stray chunk of white space on line 5013 that might as well go away now.

File: src/EnergyPlus/PackagedTerminalHeatPump.cc
Comment (Myoldmopar) [inline src/EnergyPlus/PackagedTerminalHeatPump.cc]: Same comment, toss these away if they aren't needed.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc
Comment (Myoldmopar) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:9677]: The changes in this file are just another reminder of how many _gotchas_ (potential future bugs) there in in the codebase.  Your code is fine @lgu1234, I'm just lamenting for a moment :laughing:

**Review by Myoldmopar — CHANGES_REQUESTED** [review]
There are some minor changes, code cleanup mostly.  I had a question regarding the need for a transition rule since you manually went in and updated the IDFs.  I also made a comment about the degree symbol in the IDFs that we need to resolve.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc
Comment (Nigusse) [reply src/EnergyPlus/HVACVariableRefrigerantFlow.cc]: Yes, the commented out section should be removed.

**Review by Nigusse — COMMENTED** [review]
@Myoldmopar Thank you for the review. I am addressing your comments.

**Comment by Myoldmopar** [comment]
OK, so a few things:
 - I merged develop in here, so pull before doing anything else
 - There are build warnings, just unused variables, but I didn't resolve those yet
 - I was about to start making the transition program, and found I hadn't started it for 8.8 yet, so I am going to go do that in a separate PR and then I'll merge it into this branch when ready
 - But now for the real concern.  We are actually OK with forcing a file that wasn't autosized before into being autosize based on them having a blank or zero in this field?  (https://github.com/NREL/EnergyPlus/pull/6117/files#diff-84d66175020363cd70e1e9161a145675R11)  I feel very uneasy about doing that.  What if the user doesn't have design/sizing objects in their file, won't this be a big annoyance?  Maybe I'm missing something, please clarify.

**Comment by Nigusse** [comment]
@Myoldmopar I did have the same concern as yours but forgot to mention it. Probably the best approach is not to provide a transition. Let the user file run with whatever values specified in those fields.

**Comment by Myoldmopar** [comment]
I agree.  @mjwitte ?

**Comment by mjwitte** [comment]
Not sure yet - I need some time to look this over and understand what the changes are.  BTW - this branch is conflicted now.

**Comment by Myoldmopar** [comment]
Yeah the conflict will go away if we decide we don't actually need the transition rule anyway.

**Comment by mjwitte** [comment]
@Nigusse @Myoldmopar The fundamental question is which device should be controlling the outdoor air flow.  In the original implementation, the DOAS controlled the flow to the zone unit (even though it divided up the flow incorrectly).  This fix flips that around (if I'm following correctly) so now the zone unit sets the flow rate on the OA inlet node which is one of the mixer nodes.  

* If we don't transition these to autosize, then we end up with zero OA flow and the DOAS doesn't run?
* What happens if the DOAS can't provide all the requested airflow?  Max iterations exceeded?  Or just starve the OA flow?  If it's starved, then what to the flows in the zone unit look like - just a smaller OA fraction?  
* What happens when the DOAS air handler is off but the zone unit is running?
* What happens when the DOAS air handler is on but the zone unit is off?  In a real system, for the outlet mixer, the OA would go into the zone.  For the inlet mixer, the OA proably would go to the zone?  But with the current code, I'm guessing there is no flow at the mixer if the zone unit is off.  Which may simply shut off the DOAS (whether that's the user's intent or not) or go into max iteration land.
* There is a pending new feature request for the OA flow rate to be controllable for this configuration (DCV, time-of-day, etc.).  That may influence how we do this fix.

One option to consider is to add a new input to the AirTerminal:SingleDuct:Mixer, "Design Specification Outdoor Air Object Name".  That would allow the mixer to set the OA flow rate (fixed for now, but variable with the above new feature).  This would be similar to what's done with AirTerminal:DualDuct:VAV:OutdoorAir.  But if we do this, then the zone units need to do the right thing with the OA flow (even when they are off).

File: testfiles/DOAToPTAC.idf
Comment (mjwitte) [inline testfiles/DOAToPTAC.idf:1913]: @Nigusse Curious why this change and others like it?

Comment (Nigusse) [reply testfiles/DOAToPTAC.idf:1913]: Because of changes in the amount of outdoor air (primary air), there were warning error messages that are related to capacity to flow ratio out of range.

**Comment by Nigusse** [comment]
I agree that now the OA amount is set by the Zone equipment and will be the design OA flow. It is only the zone equipment that has the information about the amount of OA. And the OA amount can be limited by the supply air flow of the parent object in the case of inlet side mixer or will be the maximum amount if it is supply side mixer.

(1) Yes it can result to zero outdoor air flow if we do not transition when the flow rates are set to zero. But the flip side is that if we transition to autosize and zone sizing is set to No, then we will have a different issue. Probably the best option is to provide only warning message during transition.

(2) The model does not control the OA amount directly, but it will be proportional to the supply air or will be the maximum design amount. The terminal mixer does not do any control except just calculating the mixed air properties based on the flow rates set by the zone equipment.

(3) I believe that if a zone equipment is off, then the terminal air mixer primary air flow rate is set to zero as well. If all zone equipment connected to DOAS are off, then there will not be any flow set to DOAS. I agree this may cause max iteration issue but I have not tested it.

(4) We set a plan to expand the original implementation to other Zone Equipment. I learned about OA control issues when I did Unit Ventilator because I had to allow controlling the OA amount for different OA control options available.

**Comment by mjwitte** [comment]
@Nigusse Thanks for the responses.  Then I'm OK with removing the transition rule for this

*But* I would like to see a warning from the zone equipment if it is connected to a mixer (ATMixerExists is true) and the design OA flow rate is zero.

**Comment by Nigusse** [comment]
@mjwitte @Myoldmopar Will add a warning message in the get input routine for each of the zone HVAC equipment connected to DOAS.

**Comment by Nigusse** [comment]
@mjwitte @Myoldmopar Below is a sample warning error messege when ATMixerExists is true and the design OA flow rate is zero.

   ** Warning ** GetPTUnit: ZoneHVAC:PackagedTerminalAirConditioner="SPACE1-1 PTAC" input value to Cooling Outdoor Air Flow Rate = 0.0 
   **   ~~~   ** Cooling Outdoor Air Flow Rate must be > 0 or set to autosize when ZoneHVAC:PackagedTerminalAirConditioner is connected to DOAS. 
   ** Warning ** GetPTUnit: ZoneHVAC:PackagedTerminalAirConditioner="SPACE1-1 PTAC" input value to Heating Outdoor Air Flow Rate = 0.0 
   **   ~~~   ** Heating Outdoor Air Flow Rate must be > 0 or set to autosize when ZoneHVAC:PackagedTerminalAirConditioner is connected to DOAS. 
   ** Warning ** GetPTUnit: ZoneHVAC:PackagedTerminalAirConditioner="SPACE1-1 PTAC" input value to No Load Outdoor Air Flow Rate = 0.0 
   **   ~~~   ** No Load Outdoor Air Flow Rate must be > 0 or set to autosize when ZoneHVAC:PackagedTerminalAirConditioner is connected to DOAS.

**Comment by mjwitte** [comment]
@Nigusse The warning is a good start.  I think it needs to reference the mixer object name (assuming that's available, or at least the mixer object type).  Something like:
** Warning ** GetPTUnit: ZoneHVAC:PackagedTerminalAirConditioner="SPACE1-1 PTAC" input value to Cooling Outdoor Air Flow Rate = 0.0
** ~~~ ** and is connected to  AirTerminal:SingleDuct:Mixer = "SPACE1-1 DOAS MIXER".  Cooling Outdoor Air Flow Rate must be > 0 or set to autosize to get flow from the DOAS.

Maybe you can word this better.

**Comment by Nigusse** [comment]
Added warning message Zone HVAC equipment: PTAC, PTHP, WSHP, VRF, and FanCoil when the OA amount is set to zero and terminal air mixer exists.  Resolved the merge conflict and did some cleanup. This is ready for review.

**Comment by mjwitte** [comment]
@Nigusse @Myoldmopar Before we merge this, I'm going to reopen some discussion on who sets the OA flow rate here - sorry.  

Prior to this fix, the air loop connected to the AirTerminal:SingleDuct:Mixer set the OA flow rate at each mixer based on the total system flow divided by the number of mixers.

This fix flips things around so that the OA flow rate set by the zone equipment (e.g. fan coil unit) is the OA flow rate at the AirTerminal:SingleDuct:Mixer object, and the airloop flow will be the sum of the mixer flows.

There is other work in discussion which proposes to add variable OA flow controls for this system configuration.  This raises the question of whether the mixer should be the one controlling the OA flow, instead of the zone equipment.

**Comment by Myoldmopar** [comment]
Alright.  @mjwitte given that there is other work in discussion, would it be unfortunately best to not make this change now?  So we avoid going in one direction just to flip-flop back the next release?  There are still a few weeks before IO freeze, so we do have time to chat.  Maybe our planning meeting this week will help clarify what's coming in OA land.  Maybe we need to utilize the sizing call this week to really dive in and decide?

If we do choose to get this in, it's conflicted yet again, and this time it's not just the transition rules file, but something in the code itself.  There's also a build warning for a simple unused parameter that should be cleaned up.

**Comment by mjwitte** [comment]
@Nigusse @lgu1234 Does #6223 make this PR obsolete?  If not, which one should drop in first?

**Comment by Nigusse** [comment]
@mjwitte  @lgu1234 Yes, this PR is going to be obsolete, no need to update it. Most of the changes were flipping inputs data from the air loop to zone equipment. Gu has moved any relevant stuff to #6223.

**Comment by mjwitte** [comment]
Superseded by #6223.

**Comment by rraustad** [comment]
What's the $ value of a rejected effort?

**Pull request closed without merging**
